packages/openai-compatible/src/openai-compatible-provider.ts

```diff
@@ -0,0 +1,193 @@
+import {
+  EmbeddingModelV1,
+  LanguageModelV1,
+  ProviderV1,
+} from '@ai-sdk/provider';
+import {
+  FetchFunction,
+  loadApiKey,
+  withoutTrailingSlash,
+} from '@ai-sdk/provider-utils';
+import {
+  OpenAICompatibleChatConfig,
+  OpenAICompatibleChatLanguageModel,
+} from './openai-compatible-chat-language-model';
+import { OpenAICompatibleChatSettings } from './openai-compatible-chat-settings';
+import { OpenAICompatibleCompletionLanguageModel } from './openai-compatible-completion-language-model';
+import { OpenAICompatibleCompletionSettings } from './openai-compatible-completion-settings';
+import { OpenAICompatibleEmbeddingSettings } from './openai-compatible-embedding-settings';
+import { OpenAICompatibleEmbeddingModel } from './openai-compatible-embedding-model';
+
+export interface OpenAICompatibleProvider<
+  CHAT_MODEL_IDS extends string = string,
+  COMPLETION_MODEL_IDS extends string = string,
+  EMBEDDING_MODEL_IDS extends string = string,
+> extends ProviderV1 {
+  (
+    modelId: CHAT_MODEL_IDS,
+    settings?: OpenAICompatibleChatSettings,
+  ): LanguageModelV1;
+
+  languageModel(
+    modelId: CHAT_MODEL_IDS,
+    settings?: OpenAICompatibleChatSettings,
+    config?: OpenAICompatibleChatConfig,
+  ): LanguageModelV1;
+
+  chatModel(
+    modelId: CHAT_MODEL_IDS,
+    settings?: OpenAICompatibleChatSettings,
+    config?: OpenAICompatibleChatConfig,
```

Comment:
Using config here (it was absent in previous iterations) means that in concrete impls e.g. TogetherAIProvider we end up having to deal with the config's [required parameters](https://github.com/vercel/ai/blob/90b9c52bea5cd6dc3c454fb70ca976d31112bc43/packages/openai-compatible/src/openai-compatible-completion-language-model.ts#L30), which would otherwise be hidden from the concrete impl (we deal with them in `getCommonModelConfig` below).

There are various ways to address, for example:
- create and use a different type here that only has what we do want to expose to the concrete impl (defaultObjectGenerationMode)
- put the bits we want to expose in settings instead, and maybe merge that into the config in `chatModel` impl logic
- make the config params optional instead of required and add validation logic elsewhere if needed
- add a method in `OpenAICompatibleProvider` or exported as part of this file that allows the concrete impls to construct a config, with args being only the bits we need from the concrete impl (but then we may want to change how the Provider is instantiated since currently we do it as part of that flow, and then the Provider pattern is more deviant from our others, which wouldn't be the end of the world but is nice to avoid)

I may also not be thinking of something, or I may have missed what you have in mind, in any case helpful to hear your design view before I go farther.

Reply:
not a fan of the 3rd arg bc
- this api is user-facing and adding args will limit our future options and impact our maintainance.
- we dont control what we expose (full config)
- config vs settings separation becomes meaningless (settings is intended for user, config internally)
- I'm trying to avoid > 2 args wherever i can. ideally 1 arg with object, in some cases (like this) 2 args for convenience. 3 gets confusing and harder to maintain

if users need to override some of the configs from the default, we can add them to `settings` (in addition to `config`).

if we want this to be extensible, we should look at some other mechanism. Fwiw I think the openai-compatible providers should fully re-implement the provider file to do what they need to do and to have max flexibility at that level (e.g. for their own API key env params, to limit the kinds of models, etc.). 

My approach would be not to restrict this part (i.e. `OpenAICompatibleProvider` should only be for direct instantiation, vanity providers should instead implement `ABCProvider` from scratch by copying and adjust this file)

Reply:
Thanks, this is super helpful and clarifies things nicely.